File: src/collectors/codex.py
"""Codex CLI collectors."""

import os
from typing import Dict
from .base import EphemeralCollector, PersistentCollector
from ..utils.tmux import EphemeralSession, PersistentSession
from ..parsers.codex import parse_codex_output


class CodexEphemeralCollector(EphemeralCollector):
    """Ephemeral collector for Codex CLI (single-shot usage)."""

    def collect(self) -> Dict[str, Dict]:
        """Collect Codex usage metrics using ephemeral tmux session.

        Returns:
            Dict with keys: '5h', 'weekly'
        """
        session_name = f"codex-usage-{os.getpid()}"

        with EphemeralSession(session_name, "codex") as session:
            # Send /status command character-by-character
            session.send_keys("/status", delay=0.2)

            # Send Enter
            session.send_keys("Enter", literal=True)

            # Poll until status output appears
            output = session.wait_for_content("limit:", timeout=8.0)

        # Parse and return metrics
        return parse_codex_output(output)


class CodexPersistentCollector(PersistentCollector):
    """Persistent collector for Codex CLI (live dashboard with session reuse)."""

    def __init__(self):
        self.session_name = f"codex-live-{os.getpid()}"
        self.session = PersistentSession(self.session_name, "codex")
        self._last_good = None  # Cache last successful metrics

    def start(self) -> Dict[str, Dict]:
        """Create persistent session and collect initial metrics.

        Windup phase: Create session → execute /status → poll for response

        Returns:
            Dict with keys: '5h', 'weekly'
        """
        # Create session and wait for prompt
        self.session.windup()

        # Execute /status command
        self.session.send_keys("/status", delay=0.2)
        self.session.send_keys("Enter", literal=True)

        # Poll until status output appears
        output = self.session.wait_for_content("limit:", timeout=8.0)

        metrics = parse_codex_output(output)
        if self._has_real_data(metrics):
            self._last_good = metrics
        return metrics

    def _has_real_data(self, metrics: Dict[str, Dict]) -> bool:
        """Check if metrics contain real parsed data (not all None/fallback)."""
        for key in ("5h", "weekly"):
            entry = metrics.get(key, {})
            if entry.get("percent_used") is not None:
                return True
        return False

    def refresh(self) -> Dict[str, Dict]:
        """Refresh metrics from existing session.

        Recurrent polling phase:
        1. Re-enter /status command
        2. Poll for response

        Returns:
            Dict with keys: '5h', 'weekly'
        """
        # Execute /status command (no need to exit previous command)
        self.session.send_keys("/status", delay=0.2)
        self.session.send_keys("Enter", literal=True)

        # Poll until status output appears
        output = self.session.wait_for_content("limit:", timeout=8.0)

        metrics = parse_codex_output(output)
        if self._has_real_data(metrics):
            self._last_good = metrics
            return metrics

        # If capture failed, return last known good metrics
        if self._last_good is not None:
            return self._last_good

        return metrics

    def stop(self):
        """Stop persistent session and cleanup."""
        self.session.winddown()

File: packages/core/src/collectors/codex.ts
/**
 * Codex CLI collectors.
 * Port of src/collectors/codex.py
 */

import type { MetricsDict, EphemeralCollector, PersistentCollector } from "../types.js";
import { EphemeralSession, PersistentSession } from "../utils/tmux.js";
import { parseCodexOutput } from "../parsers/codex.js";

/** Ephemeral collector for Codex CLI (single-shot usage) */
export class CodexEphemeralCollector implements EphemeralCollector {
  async collect(): Promise<MetricsDict> {
    const sessionName = `codex-usage-${process.pid}`;
    const session = new EphemeralSession(sessionName, "codex");

    try {
      await session.enter();

      // Send /status command character-by-character
      await session.sendKeys("/status", 200, false);

      // Send Enter
      await session.sendKeys("Enter", 0, true);

      // Poll until status output appears
      const output = await session.waitForContent("limit:", 8000);

      return parseCodexOutput(output);
    } finally {
      await session.cleanup();
    }
  }
}

/** Persistent collector for Codex CLI (live dashboard with session reuse) */
export class CodexPersistentCollector implements PersistentCollector {
  private sessionName: string;
  private session: PersistentSession;
  private _lastGood: MetricsDict | null = null;

  constructor() {
    this.sessionName = `codex-live-${process.pid}`;
    this.session = new PersistentSession(this.sessionName, "codex");
  }

  async start(): Promise<MetricsDict> {
    // Create session and wait for prompt
    await this.session.windup();

    // Execute /status command
    await this.session.sendKeys("/status", 200, false);
    await this.session.sendKeys("Enter", 0, true);

    // Poll until status output appears
    const output = await this.session.waitForContent("limit:", 8000);

    const metrics = parseCodexOutput(output);
    if (this._hasRealData(metrics)) {
      this._lastGood = metrics;
    }

    return metrics;
  }

  private _hasRealData(metrics: MetricsDict): boolean {
    for (const key of ["5h", "weekly"]) {
      const entry = metrics[key];
      if (entry && typeof entry === "object" && "used_pct" in entry && typeof entry.used_pct === "number") {
        return true;
      }
    }
    return false;
  }

  async refresh(): Promise<MetricsDict> {
    // Execute /status command
    await this.session.sendKeys("/status", 200, false);
    await this.session.sendKeys("Enter", 0, true);

    // Poll until status output appears
    const output = await this.session.waitForContent("limit:", 8000);

    const metrics = parseCodexOutput(output);
    if (this._hasRealData(metrics)) {
      this._lastGood = metrics;
      return metrics;
    }

    // If capture failed, return last known good metrics
    if (this._lastGood !== null) {
      return this._lastGood;
    }

    return metrics;
  }

  async stop(): Promise<void> {
    await this.session.winddown();
  }
}
